Split feature flags and feature previews into their own data models

gherceg opened the pull request:
## Summary
<!--
    Provide a link to the ticket or document which prompted this change,
    Describe the rationale and design decisions.
-->
[JIRA Ticket](https://dimagi-dev.atlassian.net/browse/SS-102)

Only superusers have access to feature flags, whereas anyone can access feature previews. These should be in their own respective data models with adequate permissions around the feature flags data model. 

This turned out to be a bigger PR than expected because I realized `ignore_models` was not being used everywhere it should have been. So in addition to adding superuser checks when building linked data view models, I improved the `ignore_models` behavior in the same area. As mentioned under automated tests, I also added more tests around methods that were previously lacking test coverage.

This includes a migration to add `previews` as another choice for a value in the `model` attribute of `DomainLinkHistory`, and then copy existing entries with model='toggles' into a new object with model='previews'. There are 45392 total domain link history objects, and 954 objects with model='toggles'.  

**Reviewing commit by commit might be helpful, but some are still larger than I would have liked**
## Feature Flag
<!-- If this is specific to a feature flag, which one? -->
`LINKED_DOMAINS`
## Product Description
<!-- For non-invisible changes, describe user-facing effects. -->
For users who manage linked projects and specifically rely on the Linked Project History page, previous entries that showed up as "Feature Flags and Previews" will now show up under "Feature Flags". There isn't a straightforward way to determine if past pushes/pulls were syncing just feature flags, just feature previews, or both. For this reason, it seems safest to default to every previous push/pull being containing feature flags, and also makes the migration as straightforward as possible. Moving forward, the data model the user syncs (Feature Flags vs Feature Previews) will show up accurately under the Linked Project History report. 
## Safety Assurance

- [x] Risk label is set correctly
- [x] All migrations are backwards compatible and won't block deploy
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change
- [x] If QA is part of the safety story, the "Awaiting QA" label is used
- [x] I have confidence that this PR will not introduce a regression for the reasons below

### Automated test coverage

<!-- Identify the related test coverage and the tests it would catch -->
Added tests in `test_view_helpers.py` to ensure superuser permissions are respected. Also added test coverage that was missing around `build_view_models_from_data_models` and `build_pullable_view_models`.
### QA Plan

<!--
- Describe QA plan that along with automated test coverages proves this PR is regression free
- Link to QA Ticket
-->
https://dimagi-dev.atlassian.net/browse/QA-3156
### Safety story
<!--
Describe any other pieces to the safety story including
local testing, why the change is inherently safe, and/or plans to limit the blast radius of a defect.
-->
Locally tested, and I feel good about the automated test coverage around this. 
### Rollback instructions

<!--
If this PR follows standards of revertability, check the box below.
Otherwise replace it with detailed instructions or reasons a rollback is impossible.
-->

- [x] This PR can be reverted after deploy with no further considerations

gherceg committed:
separate data models; separate update methods; add tests

gherceg committed:
handle superuser data models appropriately

gherceg committed:
hide superuser data models in history report

gherceg committed:
update references to ALL_LINKED_MODELS

gherceg committed:
add DomainLinkHistory migration to add previews as choice

gherceg committed:
isort on views.py

gherceg commented on corehq/apps/linked_domain/views.py:
```diff
@@ -238,12 +241,15 @@ def page_context(self):
         master_reports, linked_reports = get_reports(self.domain)
         master_keywords, linked_keywords = get_keywords(self.domain)
 
+        is_superuser = self.request.couch_user.is_superuser
```
Is this the best way to obtain whether user is a superuser?

orangejenny replied:
yep

gherceg commented on corehq/apps/linked_domain/views.py:
```diff
@@ -387,6 +389,11 @@ def total_records(self):
         query = DomainLinkHistory.objects.filter(link=self.selected_link)
+
+        # filter out superuser data models
+        if not self.request.couch_user.is_superuser:
```
Same here.

gherceg committed:
duplicate toggle events into preview events

gherceg commented on corehq/apps/linked_domain/migrations/0015_add_previews_to_choices_20210617_2010.py:
```diff
@@ -0,0 +1,47 @@
+from corehq.apps.linked_domain.models import DomainLinkHistory
+
+
+def _duplicate_toggle_events_into_preview_events(apps, schema_editor):
+    all_toggle_events = DomainLinkHistory.objects.filter(model=MODEL_FLAGS)
+    # iterate through events, copy into a new entry with model='previews'
+    for toggle_event in all_toggle_events:
+        toggle_event.pk = None
+        toggle_event.model = MODEL_PREVIEWS
+        toggle_event.save()
```
Want to bring extra attention here. Is this the best way to go about this?

orangejenny replied:
yep, this is good

biyeun commented on corehq/apps/linked_domain/local_accessors.py:
```diff
@@ -12,13 +12,21 @@
     }
 
 
+def get_enabled_toggles(domain):
```
nice ^_^

orangejenny commented on corehq/apps/linked_domain/const.py:
```diff
@@ -45,6 +45,7 @@
     (MODEL_HMAC_CALLOUT_SETTINGS, ugettext_lazy('Signed Callout')),
 ]
 
+NON_SUPERUSER_DATA_MODELS = INDIVIDUAL_DATA_MODELS + DOMAIN_LEVEL_DATA_MODELS + FEATURE_FLAG_DATA_MODELS
```
I'm not sold this constant is pulling its weight. Looks like there's just one usage that could be replaced by a `set(LINKED_MODELS) - set(SUPERUSER_DATA_MODELS)`

gherceg replied:
I respect that. Removed it for now.

orangejenny commented on corehq/apps/linked_domain/README.md:
```diff
@@ -36,15 +36,16 @@ supported applications. Remote domain linkages cannot be created via the UI; see
 
 Linked domains share configuration data. Supported data types are defined in
-[LINKED_MODELS](https://github.com/dimagi/commcare-hq/blob/966b62cc113b56af771906def76833446b4ba025/corehq/apps/linked_domain/const.py#L13):
+[ALL_LINKED_MODELS](https://github.com/dimagi/commcare-hq/blob/master/corehq/apps/linked_domain/const.py#L43):
```
It's a good catch to update this, but the line number is likely to get out of date. Maybe instead of a link, this doc could just say `corehq.apps.linked_domain.const.ALL_LINKED_MODELS` or even just `ALL_LINKED_MODELS`? It should be easy enough for readers to grep for the definition.

orangejenny commented on corehq/apps/linked_domain/README.md:
```diff
@@ -36,15 +36,16 @@ supported applications. Remote domain linkages cannot be created via the UI; see
 - Keywords
 - User roles
 - Custom data fields for users, products, and locations
-- Feature flags and privileges
```
...oops, "privileges" was a lie.

orangejenny commented on corehq/apps/linked_domain/migrations/0015_add_previews_to_choices_20210617_2010.py:
```diff
@@ -2,6 +2,18 @@
+from corehq.apps.linked_domain.models import DomainLinkHistory
+
+
+def _duplicate_toggle_events_into_preview_events(apps, schema_editor):
```
Might as well `@skip_on_fresh_install` this function.

orangejenny reviewed (approved):
My comments are all minor.

gherceg commented on corehq/apps/linked_domain/updates.py:
```diff
@@ -112,25 +116,42 @@ def update_model_type(domain_link, model_type, model_detail=None):
+    # enable downstream toggles that are enabled upstream
+    _set_toggles(upstream_toggles - downstream_toggles, True)
+    # disable downstream toggles that are disabled upstream
+    _set_toggles(downstream_toggles - upstream_toggles, False)
```
Yeah this is new behavior and no they have not been given the chance to sign off. What is the most efficient way to go about that?

orangejenny replied:
There's a cross-divisional product team meeting each week, you can ask Kirti to have it discussed there.

gherceg commented on corehq/apps/linked_domain/tests/test_update_previews.py:
```diff
@@ -0,0 +1,43 @@
+)
+
+
+@mock.patch('corehq.feature_previews.all_previews_by_name', mock.Mock(return_value={'TEST_PREVIEW': TEST_PREVIEW}))
```
I did this for feature flags as well. I agree that it would be better to avoid mocks, but I also didn't want to depend on a random production feature flag or preview in the event that it changes in the future. For example, I don't like the idea of removing a feature preview resulting in this test failing. 

Ideally I could have created a toggle and registered it in a way that `all_previews_by_name` finds it, but was unable to do so. If you have any ideas I'm open to hearing them.

gherceg commented on corehq/apps/linked_domain/tests/test_update_toggles.py:
```diff
@@ -0,0 +1,43 @@
+
+    def test_get_enabled_toggles(self):
+        set_toggle('test_toggle', self.domain_link.master_domain, True, NAMESPACE_DOMAIN)
+        set_toggle('report_builder', self.domain_link.master_domain, True, NAMESPACE_DOMAIN)
```
Huh I forgot to remove this line. Curious that this test passes. Looking into it now.

gherceg replied:
Oh I'm mocking the method nevermind. Exhibit A of why mocking things isn't ideal. Couldn't help but hear Dumbledore saying "Actually if I think about it, it doesn't seem curious at all."

gherceg committed:
cleanup tests

gherceg commented on corehq/apps/linked_domain/tests/test_update_previews.py:
```diff
@@ -0,0 +1,43 @@
+        self.assertTrue(TEST_PREVIEW.enabled(self.domain_link.linked_domain, NAMESPACE_DOMAIN))
+
+    def test_downstream_preview_disabled_in_local_link(self):
+        set_toggle('test_preview', self.domain_link.master_domain, True, NAMESPACE_DOMAIN)
```
Yeah thanks for flagging. Looking at it again, the intention was just to ensure the downstream domain had the flag enabled, so I can just rely on the `set_toggle` method and clean this up that way.

gherceg committed:
address Jenny's feedback

- remove link to too specific of a line number
- remove NON_SUPERUSER_DATA_MODELS
- add skip on fresh install to migration

gherceg commented:
Sorry for the delay. Addressed @orangejenny's feedback and will put on staging for QA if tests pass. The feature flag behavior change on push/pull should be brought up in this Friday's product meeting.

gherceg committed:
Merge branch 'master' into gh/split-toggles-and-previews

gherceg committed:
do not disable downstream flags if disabled upstream

gherceg committed:
do not disable downstream previews if disabled upstream

gherceg commented:
There was a concern with disabling toggles/previews downstream if disabled upstream, so I reverted those changes which are backed up by tests. This passed QA prior to this change but I am confident this still functions appropriately.
